File: docs/chapters/02-design-patterns/java/PortsAdaptersInventory.java
// Scenario: reserve stock via modern WMS or legacy warehouse API
// Demonstrates: Port in domain + Adapter in infrastructure (hexagonal)
// Trade-off: mapping layer vs leaking vendor DTOs into OrderService

interface InventoryPort {
    boolean reserve(String sku, int qty);
}

final class OrderService {
    private final InventoryPort inventory;
    OrderService(InventoryPort inventory) { this.inventory = inventory; }
    boolean place(String sku, int qty) { return inventory.reserve(sku, qty); }
}

final class LegacyWarehouseAdapter implements InventoryPort {
    public boolean reserve(String sku, int qty) {
        // translate to legacy SOAP/JSON and map errors to boolean or domain error
        return true;
    }
}

File: docs/chapters/02-design-patterns/go/ports_adapters_inventory.go
// Scenario: reserve stock via modern WMS or legacy warehouse API
// Demonstrates: Port in domain + Adapter in infrastructure
// Trade-off: explicit error mapping at boundary vs bool-only port

package main

type InventoryPort interface {
	Reserve(sku string, qty int) error
}

type OrderService struct{ inventory InventoryPort }

func (s *OrderService) Place(sku string, qty int) error {
	return s.inventory.Reserve(sku, qty)
}

type LegacyWarehouseAdapter struct{}

func (LegacyWarehouseAdapter) Reserve(string, int) error {
	// call legacy API; map vendor errors to domain errors here
	return nil
}
